Assorted changes 7

leventov opened the pull request:
1. Extract common `BaseWritableMemoryImpl` and `BaseWritableBufferImpl` in preparation for introduction of non-native Memory and Buffer impls.

2. I cannot agree with the current state of copyTo().

 - You previously said that you don't see long pauses connected to Unsafe.getMemory() anymore. Checking the OpenJDK development history proves that it was indeed fixed, but only in OpenJDK 9, not in OpenJDK 8. So we need to retain chunked copy logic.
 - I made `copyTo()` to tolerate overlapping ranges, but not when the ranges match exactly, because I don't see any cases (including "corner") when it could possibly be a benign intention of the caller, not a bug.
 - I added checks (and chunked copy) to all "getArray" and "putArray" methods. I think the library should be safe and correct first, and fast only after that. Let's not leave a gate for potentially very hard-to-find bugs, until it proves to be a bottleneck in applications. However, I will agree to put those extra checks under `assert`, if you want.

leventov committed:
Extract byte order agnostic BaseWritableMemoryImpl

leventov committed:
Remove unused field

leventov committed:
Extract byte order agnostic BaseWritableBufferImpl

leventov committed:
Remove unused field in WritableBufferImpl

leventov committed:
Overlap-checking and chunked Memory.copyTo(); safer getArray and putArray methods

Commit:
Code style issues:

Missing finals. Missing space before javadoc.
Wrapped operators should be at beginning of next line, not at the end of
a line.

leventov commented on src/main/java/com/yahoo/memory/BaseWritableMemoryImpl.java:
```diff
@@ -168,21 +169,21 @@ private static void copyMemorySlowPath(final Object srcUnsafeObj, final long src
   }
 
-  /** @see #UNSAFE_COPY_MEMORY_THRESHOLD */
+  /* @see #UNSAFE_COPY_MEMORY_THRESHOLD */
```
Why turning a javadoc comment into a non-javadoc comment?

leerho replied:
1) As a javadoc comment it was very incomplete.  It was missing documentation of all 5 parameters. Because of this it raised 5 warnings in my IDE.  You need to have similar detections set up in your IDE.
2) The @see is a reference, which should be used as additional information after an english sentence that explains the intent of the method. The @see did not explain the intent of the method, it is more of a code comment than a javadoc comment.
3) It is a private method.  Javadocs are always good, but not required for private methods.  If you want to write a javadoc, then make sure it is at least a complete javadoc.

leventov replied:
I made this a Javadoc method to enable easy navigation in the IDE (Ctrl + click). The method and the constant have the common description, so I put it on the constant's Javadoc, and the method's one just links to it. It could have been vice versa.

Yes, because this method is private, all those strict Javadoc lint rules shouldn't necessarily be followed. It's purpose just to easy reading, understanding, navigation for the code developers.

leventov replied:
But it's not that important, it was just a question. Code could stay the same

leerho replied:
I don't know how to set more relaxed lint rules for private methods.  In my eclipse and checkstyle, a javadoc is not required for private methods.  But once you start a javadoc with /** all the lint rules apply.  

When I first open a piece of code, one of the first things I do is check all the warnings.  If they are easy to fix, I usually just fix them and go on.  This one, however, was harder to fix properly, and I didn't think that changing it to a code comment lost any information, since private javadocs don't appear in the public javadocs, the only  way someone would see this would be when they are looking at the code.

leventov committed:
Move methods up

Commit:
Removed unused imports

leventov committed:
Fix a bug in Memory.compareTo() when comparing ranges with same start but different lengths; Refine Memory.compareTo() and Buffer.compareTo() javadocs

leerho commented on src/main/java/com/yahoo/memory/BaseWritableBufferImpl.java:
```diff
@@ -111,16 +111,17 @@ public void getByteArray(final byte[] dstArray, final int dstOffset, final int l
     final long thatAdd = that.getCumulativeOffset() + thatOffsetBytes;
-    final Object thisObj = (isDirect()) ? null : unsafeObj;
-    final Object thatObj = (that.isDirect()) ? null : ((WritableBuffer)that).getArray();
+    @SuppressWarnings("UnnecessaryLocalVariable") final Object thisObj = this.unsafeObj;
```
My IDE declares that these @SuppressWarnings("UnnecessaryLocalVariable") are unsupported.  Never seen them before.  And I don't agree with warning's claim that these vars are unnecessary.  Dereferencing these before the loop makes all kinds of sense and we can't rely on the compiler or JIT to do it.

leventov committed:
Remove annotations

leerho commented on src/main/java/com/yahoo/memory/BaseWritableMemoryImpl.java:
```diff
@@ -0,0 +1,416 @@
+    // Time to Safe Point pauses.
+    unsafe.copyMemory(srcUnsafeObj, srcAdd, dstUnsafeObj, dstAdd, lengthBytes);
+  }
+
```
This `copyMemoryCheckingNonOverlapping(...)` is only used for `put/get<primitive>Array(...)`.  
I don't understand the need.  The only way an overlap could occur would be if someone tried to use one of these methods to get or put an array into the same heap array that they wrapped.  I.e.

```java
    byte[] srcAndDst = new byte[128];
    WritableMemory wmem = WritableMemory.wrap(srcAndDst);
    wmem.getByteArray(0, srcAndDst, 64, 64);  //non-overlapping
    wmem.getByteArray(0, srcAndDst, 32, 64); //overlapping
```
This makes no sense to me, the user should use `copyTo()`!
We should check that the dst or src array is not the same as _this.getArray()_, otherwise throw an error. 
Much simpler.

leventov replied:
Ok.

leerho replied:
Can you fix?

leventov replied:
Fixed.

Commit:
Add unit tests

leventov committed:
Less checks in getArray/putArray

Commit:
Operators belong on the next line.

Commit:
Fixed bug. Created common class for all compare and copy logic.

BaseWritableBufferImpl had a bug: missing if ((thisObj != thatObj) ||
(thisAdd != thatAdd)) {...}.
Added javadocs and code comments.
Added unit tests.

leventov commented on src/main/java/com/yahoo/memory/CompareAndCopy.java:
```diff
@@ -0,0 +1,134 @@
+   */
+  private static void copyMemoryOverlapAddressCheck(final Object srcUnsafeObj, final long srcAdd,
+      final Object dstUnsafeObj, final long dstAdd, final long lengthBytes) {
+    if (((srcAdd + lengthBytes) <= dstAdd) || ((dstAdd + lengthBytes) <= srcAdd)) {
```
BTW addresses could be negative (Java's `long` is always signed), so we must use overflow-aware comparisons like `a - b <= 0` instead of `a <= b`.

leventov replied:
Or maybe not, needs to research

leerho replied:
Good catch.

leventov replied:
Sorry, it was dumb. All user-space addresses are positive. I've reverted your commit.

leventov replied:
Do you have any other concerns about this PR?

leerho replied:
Need to add equals(...) and fix the java version so that it works for 1.7 and >= 1.8.  If the user needs to add JVM args to allow illegal access, so be it, for now.

leventov replied:
It could be done in separate PRs.

Commit:
Changed to overflow-aware comparison.

Commit:
Sync with origin

Commit:
Add Memory.equalTo(Memory that) and equalTo(long, Memory, long, long);
